File: app/src/main/java/info/nightscout/androidaps/plugins/constraints/objectives/objectives/Objective.java
package info.nightscout.androidaps.plugins.constraints.objectives.objectives;

import android.app.Activity;
import android.content.Context;
import android.graphics.Color;
import android.text.util.Linkify;
import android.view.View;
import android.widget.CheckBox;
import android.widget.TextView;

import androidx.annotation.StringRes;

import java.util.ArrayList;
import java.util.List;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.utils.DateUtil;
import info.nightscout.androidaps.utils.SP;
import info.nightscout.androidaps.utils.T;

public abstract class Objective {

    private String spName;
    @StringRes
    private int objective;
    @StringRes
    private int gate;
    private long startedOn;
    private long accomplishedOn;
    List<Task> tasks = new ArrayList<>();
    public boolean hasSpecialInput = false;

    public Objective(String spName, @StringRes int objective, @StringRes int gate) {
        this.spName = spName;
        this.objective = objective;
        this.gate = gate;
        startedOn = SP.getLong("Objectives_" + spName + "_started", 0L);
        accomplishedOn = SP.getLong("Objectives_" + spName + "_accomplished", 0L);
        if ((accomplishedOn - DateUtil.now()) > T.hours(3).msecs() || (startedOn - DateUtil.now()) > T.hours(3).msecs()) { // more than 3 hours in the future
            startedOn = 0;
            accomplishedOn = 0;
        }
        setupTasks(tasks);
        for (Task task : tasks) task.objective = this;
    }

    public boolean isCompleted() {
        for (Task task : tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted())
                return false;
        }
        return true;
    }

    public boolean isCompleted(long trueTime) {
        for (Task task : tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted(trueTime))
                return false;
        }
        return true;
    }

    public boolean isRevertable() {
        return false;
    }

    public boolean isAccomplished() {
        return accomplishedOn != 0 && accomplishedOn < DateUtil.now();
    }

    public boolean isStarted() {
        return startedOn != 0;
    }

    public long getStartedOn() {
        return startedOn;
    }

    public int getObjective() {
        return objective;
    }

    public int getGate() {
        return gate;
    }

    public void setStartedOn(long startedOn) {
        this.startedOn = startedOn;
        SP.putLong("Objectives_" + spName + "_started", startedOn);
    }

    public void setAccomplishedOn(long accomplishedOn) {
        this.accomplishedOn = accomplishedOn;
        SP.putLong("Objectives_" + spName + "_accomplished", accomplishedOn);
    }

    public long getAccomplishedOn() {
        return accomplishedOn;
    }

    protected void setupTasks(List<Task> tasks) {

    }

    public List<Task> getTasks() {
        return tasks;
    }

    public void specialAction(Activity activity, String input) {}

    public abstract class Task {
        @StringRes
        private int task;
        private Objective objective;
        ArrayList<Hint> hints = new ArrayList<>();

        public Task(@StringRes int task) {
            this.task = task;
        }

        public @StringRes int getTask() {
            return task;
        }

        protected Objective getObjective() {
            return objective;
        }

        public abstract boolean isCompleted();
        public boolean isCompleted(long trueTime) { return isCompleted(); };

        public String getProgress() {
            return MainApp.gs(isCompleted() ? R.string.completed_well_done : R.string.not_completed_yet);
        }

        Task hint(Hint hint) {
            hints.add(hint);
            return this;
        }

        public ArrayList<Hint> getHints() {
            return hints;
        }

        public boolean shouldBeIgnored() {
            return false;
        }
    }

    public class MinimumDurationTask extends Task {

        private long minimumDuration;

        MinimumDurationTask(long minimumDuration) {
            super(R.string.time_elapsed);
            this.minimumDuration = minimumDuration;
        }

        @Override
        public boolean isCompleted() {
            return getObjective().isStarted() && System.currentTimeMillis() - getObjective().getStartedOn() >= minimumDuration;
        }

        @Override
        public boolean isCompleted(long trueTime) {
            return getObjective().isStarted() && trueTime - getObjective().getStartedOn() >= minimumDuration;
        }

        @Override
        public String getProgress() {
            return getDurationText(System.currentTimeMillis() - getObjective().getStartedOn())
                    + " / " + getDurationText(minimumDuration);
        }

        private String getDurationText(long duration) {
            int days = (int) Math.floor((double) duration / T.days(1).msecs());
            int hours = (int) Math.floor((double) duration / T.hours(1).msecs());
            int minutes = (int) Math.floor((double) duration / T.mins(1).msecs());
            if (days > 0) return MainApp.gq(R.plurals.objective_days, days, days);
            else if (hours > 0) return MainApp.gq(R.plurals.objective_hours, hours, hours);
            else return MainApp.gq(R.plurals.objective_minutes, minutes, minutes);
        }
    }

    public class ExamTask extends Task {
        @StringRes
        int question;
        ArrayList<Option> options = new ArrayList<>();
        private String spIdentifier;
        private boolean answered;
        private long disabledTo;

        ExamTask(@StringRes int task, @StringRes int question, String spIdentifier) {
            super(task);
            this.question = question;
            this.spIdentifier = spIdentifier;
            answered = SP.getBoolean("ExamTask_" + spIdentifier, false);
            disabledTo = SP.getLong("DisabledTo_" + spIdentifier, 0L);
        }

        public void setDisabledTo(long newState) {
            disabledTo = newState;
            SP.putLong("DisabledTo_" + spIdentifier, disabledTo);
        }

        public long getDisabledTo() {
            return disabledTo;
        }

        public boolean isEnabledAnswer() {
            return disabledTo < DateUtil.now();
        }

        public void setAnswered(boolean newState) {
            answered = newState;
            SP.putBoolean("ExamTask_" + spIdentifier, answered);
        }

        public boolean getAnswered() {
            return answered;
        }

        ExamTask option(Option option) {
            options.add(option);
            return this;
        }

        public @StringRes int getQuestion() {
            return question;
        }

        public List<Objective.Option> getOptions() {
            return options;
        }

        @Override
        public boolean isCompleted() {
            return answered;
        }
    }

    public class Option {
        @StringRes int option;
        boolean isCorrect;

        CheckBox cb; // TODO: change it, this will block releasing memeory

        Option(@StringRes int option, boolean isCorrect) {
            this.option = option;
            this.isCorrect = isCorrect;
        }

        public boolean isCorrect() {
            return isCorrect;
        }

        public CheckBox generate(Context context) {
            cb = new CheckBox(context);
            cb.setText(option);
            return cb;
        }

        public boolean evaluate() {
            boolean selection = cb.isChecked();
            if (selection && isCorrect) return true;
            if (!selection && !isCorrect) return true;
            return false;
        }
    }

    public class Hint {
        @StringRes int hint;

        Hint(@StringRes int hint) {
            this.hint = hint;
        }

        public TextView generate(Context context) {
            TextView textView = new TextView(context);
            textView.setText(hint);
            textView.setAutoLinkMask(Linkify.WEB_URLS);
            textView.setLinksClickable(true);
            textView.setLinkTextColor(Color.YELLOW);
            Linkify.addLinks(textView, Linkify.WEB_URLS);
            return textView;
        }
    }
}

File: plugins/constraints/src/main/kotlin/app/aaps/plugins/constraints/objectives/objectives/Objective.kt
package app.aaps.plugins.constraints.objectives.objectives

import android.content.Context
import android.text.util.Linkify
import android.widget.CheckBox
import android.widget.TextView
import androidx.annotation.StringRes
import app.aaps.core.data.time.T
import app.aaps.core.interfaces.resources.ResourceHelper
import app.aaps.core.interfaces.utils.DateUtil
import app.aaps.core.keys.interfaces.Preferences
import app.aaps.plugins.constraints.R
import app.aaps.plugins.constraints.objectives.keys.ObjectivesBooleanComposedKey
import app.aaps.plugins.constraints.objectives.keys.ObjectivesLongComposedKey
import kotlinx.coroutines.Runnable
import kotlin.math.floor

abstract class Objective(
    val preferences: Preferences,
    val rh: ResourceHelper,
    val dateUtil: DateUtil,
    private val spName: String,
    @StringRes val objective: Int,
    @StringRes val gate: Int
) {
    var startedOn: Long = 0
        get() = preferences.get(ObjectivesLongComposedKey.Started, spName)
        set(value) {
            field = value
            preferences.put(ObjectivesLongComposedKey.Started, spName, value = value)
        }
    var accomplishedOn: Long = 0
        get() {
            var value = preferences.get(ObjectivesLongComposedKey.Accomplished, spName)
            if (value - dateUtil.now() > T.hours(3).msecs() || startedOn - dateUtil.now() > T.hours(3).msecs()) { // more than 3 hours in the future
                startedOn = 0
                accomplishedOn = 0
                value = 0
            }
            return value
        }
        set(value) {
            field = value
            preferences.put(ObjectivesLongComposedKey.Accomplished, spName, value = value)
        }

    var tasks: MutableList<Task> = ArrayList()

    val isCompleted: Boolean
        get() {
            for (task in tasks) {
                if (!task.shouldBeIgnored() && !task.isCompleted()) return false
            }
            return true
        }

    fun isCompleted(trueTime: Long): Boolean {
        for (task in tasks) {
            if (!task.shouldBeIgnored() && !task.isCompleted(trueTime)) return false
        }
        return true
    }

    val isAccomplished: Boolean
        get() = accomplishedOn != 0L && accomplishedOn < dateUtil.now()
    val isStarted: Boolean
        get() = startedOn != 0L

    abstract inner class Task(var objective: Objective, @StringRes val task: Int) {

        var hints = ArrayList<Hint>()
        var learned = ArrayList<Learned>()

        abstract fun isCompleted(): Boolean

        open fun isCompleted(trueTime: Long): Boolean = isCompleted()

        open val progress: String
            get() = rh.gs(if (isCompleted()) R.string.completed_well_done else R.string.not_completed_yet)

        fun hint(hint: Hint): Task {
            hints.add(hint)
            return this
        }

        fun learned(learned: Learned): Task {
            this.learned.add(learned)
            return this
        }

        open fun shouldBeIgnored(): Boolean = false
    }

    inner class MinimumDurationTask internal constructor(objective: Objective, private val minimumDuration: Long) : Task(objective, R.string.time_elapsed) {

        override fun isCompleted(): Boolean =
            objective.isStarted && System.currentTimeMillis() - objective.startedOn >= minimumDuration

        override fun isCompleted(trueTime: Long): Boolean {
            return objective.isStarted && trueTime - objective.startedOn >= minimumDuration
        }

        override val progress: String
            get() = (getDurationText(System.currentTimeMillis() - objective.startedOn)
                + " / " + getDurationText(minimumDuration))

        private fun getDurationText(duration: Long): String {
            val days = floor(duration.toDouble() / T.days(1).msecs()).toInt()
            val hours = floor(duration.toDouble() / T.hours(1).msecs()).toInt()
            val minutes = floor(duration.toDouble() / T.mins(1).msecs()).toInt()
            return when {
                days > 0  -> rh.gq(app.aaps.core.ui.R.plurals.days, days, days)
                hours > 0 -> rh.gq(app.aaps.core.ui.R.plurals.hours, hours, hours)
                else      -> rh.gq(app.aaps.core.ui.R.plurals.minutes, minutes, minutes)
            }
        }
    }

    inner class UITask internal constructor(objective: Objective, @StringRes task: Int, private val spIdentifier: String, val code: (context: Context, task: UITask, callback: Runnable) -> Unit) : Task(objective, task) {

        var answered: Boolean = false
            set(value) {
                field = value
                preferences.put(ObjectivesBooleanComposedKey.AnsweredUi, spIdentifier, value = value)
            }

        init {
            answered = preferences.get(ObjectivesBooleanComposedKey.AnsweredUi, spIdentifier)
        }

        override fun isCompleted(): Boolean = answered
    }

    inner class ExamTask internal constructor(objective: Objective, @StringRes task: Int, @StringRes val question: Int, private val spIdentifier: String) : Task(objective, task) {

        var options = ArrayList<Option>()
        var answered: Boolean = false
            set(value) {
                field = value
                preferences.put(ObjectivesBooleanComposedKey.AnsweredExam, spIdentifier, value = value)
            }
        var disabledTo: Long = 0
            set(value) {
                field = value
                preferences.put(ObjectivesLongComposedKey.DisabledTo, spIdentifier, value = value)
            }

        init {
            answered = preferences.get(ObjectivesBooleanComposedKey.AnsweredExam, spIdentifier)
            disabledTo = preferences.get(ObjectivesLongComposedKey.DisabledTo, spIdentifier)
        }

        override fun isCompleted(): Boolean = answered

        fun isEnabledAnswer(): Boolean = disabledTo < dateUtil.now()

        fun option(option: Option): ExamTask {
            options.add(option)
            return this
        }
    }

    inner class Option internal constructor(@StringRes var option: Int, var isCorrect: Boolean) {

        private var cb: CheckBox? = null // TODO: change it, this will block releasing memory

        fun generate(context: Context): CheckBox {
            cb = CheckBox(context)
            cb?.setText(option)
            return cb!!
        }

        fun evaluate(): Boolean {
            val selection = cb!!.isChecked
            return if (selection && isCorrect) true else !selection && !isCorrect
        }
    }

    inner class Hint internal constructor(@StringRes var hint: Int) {

        fun generate(context: Context): TextView {
            val textView = TextView(context)
            textView.setText(hint)
            textView.autoLinkMask = Linkify.WEB_URLS
            textView.linksClickable = true
            textView.setLinkTextColor(rh.gac(context, com.google.android.material.R.attr.colorSecondary))
            Linkify.addLinks(textView, Linkify.WEB_URLS)
            return textView
        }
    }

    inner class Learned internal constructor(@StringRes var learned: Int)
}